### PR TITLE
[pycryptodome] Add fallback for `pycryptodome`

### PR DESCRIPTION
**Pull request by sulyi**
### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Implementation of AES-GCM, removes `pycryptodome` dependency

### CODE REVIEW

File: yt_dlp/downloader/hls.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class HlsFD(FragmentFD):
     FD_NAME = 'hlsnative'
 
     @staticmethod
-    def can_download(manifest, info_dict, allow_unplayable_formats=False, with_crypto=can_decrypt_frag):
+    def can_download(manifest, info_dict, allow_unplayable_formats=False, with_crypto=True):
```
Comment (sulyi) [inline yt_dlp/downloader/hls.py:32]: This is annoying, should be fixed.
I just hoping I didn't miss other usage of removed flags.

**Comment by sulyi** [comment]
Obviously `aes_ctr_encrypt` is still missing, so `aes_gcm_decrypt_and_verify` does not work yet.

File: yt_dlp/aes.py [hunk 2]
```diff
@@ -358,4 +408,76 @@ def inc(data):
...
+    block_z = [0] * BLOCK_SIZE_BYTES
+
+    for i in block_x:
+        for bit in range(7, -1, -1):
+            if i & (1 << bit):
+                block_z = xor(block_z, block_v)
+
+            do_xor = block_v[-1] & 1
+            block_v = bytes_to_intlist(
+                (int.from_bytes(intlist_to_bytes(block_v), 'big') >> 1).to_bytes(BLOCK_SIZE_BYTES, 'big')
```
Comment (sulyi) [inline yt_dlp/aes.py:444]: Is there a better way to shift the whole block?

**Comment by pukkandan** [comment]
The extractors/cookies can use the native cbc implementation without issue since they are only decrypting small sections of data. However, for the downloaders, the performance difference could be significant since large amounts of data is being decrypted. So I have changed the interface to use pycryptodome if the module is found and the native implementation otherwise.

You can overhaul the `aes` module to use bytes instead of intlist if you want as long as you keep thin wrappers for all the existing functions

Please mark the PR as draft until you have a working implementation.

**Comment by sulyi** [comment]
Compatibility could also easily handle pycryptodome and pycryptodomex differences.

**Comment by pukkandan** [comment]
@sulyi see [`9d9b2a2` (#938)](https://github.com/yt-dlp/yt-dlp/pull/938/commits/9d9b2a287633a7c5f1036e52bb8fbfb385e8dcc9)

File: yt_dlp/compat.py [hunk 3]
```diff
@@ -130,6 +130,31 @@ def compat_asyncio_run(coro):
...
+        return compat_pycrypto_AES.new(key, compat_pycrypto_AES.MODE_GCM, nonce).decrypt_and_verify(data, tag)
+else:
+    from .aes import aes_cbc_decrypt
+    from .utils import bytes_to_intlist, intlist_to_bytes
+
+    def compat_aes_cbc_decrypt(data, key, iv):
+        return intlist_to_bytes(aes_cbc_decrypt(*map(bytes_to_intlist, (data, key, iv))))
+
+    def compat_aes_gcm_decrypt_and_verify(data, key, tag, nonce):
+        return intlist_to_bytes(aes_cbc_decrypt(*map(bytes_to_intlist, (data, key, tag, nonce))))
```
Comment (pukkandan) [inline yt_dlp/compat.py:156]: Not aes_cbc_decrypt

File: yt_dlp/cookies.py [hunk 4]
```diff
@@ -655,22 +643,29 @@ def _decrypt_aes_cbc(ciphertext, key, logger, initialization_vector=b' ' * 16):
...
 def _decrypt_aes_gcm(ciphertext, key, nonce, authentication_tag, logger):
-    cipher = AES.new(key, AES.MODE_GCM, nonce)
     try:
-        plaintext = cipher.decrypt_and_verify(ciphertext, authentication_tag)
+        plaintext = aes_gcm_decrypt_and_verify(
+            bytes_to_intlist(ciphertext),
+            bytes_to_intlist(key),
+            bytes_to_intlist(authentication_tag),
+            bytes_to_intlist(nonce)
+        )
```
Comment (pukkandan) [inline yt_dlp/cookies.py:658]: u didnt replace this with its compat

File: yt_dlp/aes.py [hunk 5]
```diff
@@ -233,6 +232,8 @@ def aes_decrypt_text(data, password, key_size_bytes):
     @param {int} key_size_bytes          Possible values: 16 for 128-Bit, 24 for 192-Bit or 32 for 256-Bit
     @returns {str}                       Decrypted data
     """
+    from .compat import compat_b64decode
+
     NONCE_LENGTH_BYTES = 8
 
     data = bytes_to_intlist(compat_b64decode(data))
```
Comment (pukkandan) [inline yt_dlp/aes.py:239]: just do `base64.b64decode`

Comment (sulyi) [reply yt_dlp/aes.py:239]: I'm moving them into local import, `utils` import in `compat` as well,

File: yt_dlp/aes.py [hunk 6]
```diff
@@ -43,6 +66,10 @@ def aes_cbc_decrypt(data, key, iv):
     @param {int[]} iv          16-Byte IV
     @returns {int[]}           decrypted data
     """
+    if AES:
+        data, key, iv = map(intlist_to_bytes, (data, key, iv))
```
Comment (pukkandan) [inline yt_dlp/aes.py:70]: we are converting bytes to intlist back to bytes for no reason :/
same when returning

Comment (sulyi) [reply yt_dlp/aes.py:70]: I'm aware. I'll look into it.

File: yt_dlp/YoutubeDL.py [hunk 7]
```diff
@@ -3292,13 +3293,12 @@ def python_implementation():
...
         self._write_string('[debug] exe versions: %s\n' % exe_str)
 
-        from .downloader.fragment import can_decrypt_frag
         from .downloader.websocket import has_websockets
         from .postprocessor.embedthumbnail import has_mutagen
         from .cookies import SQLITE_AVAILABLE, KEYRING_AVAILABLE
 
         lib_str = ', '.join(sorted(filter(None, (
-            can_decrypt_frag and 'pycryptodome',
+            crypto_name and crypto_name,
```
Comment (sulyi) [inline yt_dlp/YoutubeDL.py:3301]: Please, check this!

File: yt_dlp/aes.py (hunk 6 above)
Comment (sulyi) [reply yt_dlp/aes.py:70]: Changing signature should be done consistently over the module, which is just a little bit of busy work.
My concern is, that `aes_encrypt` and `aes_ctr_decrypt` functions, are used both internally and externally, resulting the issue to be only kicked around.

**Comment by sulyi** [comment]
Handling fallback argument conversion:  https://github.com/sulyi/yt-dlp/pull/1

**Comment by sulyi** [comment]
It's still being skipped, might be due to change in GA API, but I couldn't find the definite reason.

File: yt_dlp/extractor/ivi.py [hunk 8]
```diff
@@ -96,6 +96,7 @@ def _real_extract(self, url):
         for site in (353, 183):
             content_data = (data % site).encode()
             if site == 353:
+                # TODO: move to `comapt`
```
Comment (sulyi) [inline yt_dlp/extractor/ivi.py:99]: Before it falls through the cracks, I'm making a note here as well.

File: yt_dlp/aes.py [hunk 9]
```diff
@@ -178,31 +219,62 @@ def aes_decrypt_text(data, password, key_size_bytes):
...
     """
+
     NONCE_LENGTH_BYTES = 8
 
     data = bytes_to_intlist(compat_b64decode(data))
     password = bytes_to_intlist(password.encode('utf-8'))
 
     key = password[:key_size_bytes] + [0] * (key_size_bytes - len(password))
-    key = aes_encrypt(key[:BLOCK_SIZE_BYTES], key_expansion(key)) * (key_size_bytes // BLOCK_SIZE_BYTES)
+    # FIXME: fb_functions expect expanded key
```
Comment (sulyi) [inline yt_dlp/aes.py:229]: DO NOT MERGE!

**Comment by sulyi** [comment]
This is start getting out of hand. I could write an inverse for `key_expansion`, but that's just stupid. And I won't do it.

Otherwise, I can't keep functionality the same, and provide compatibility.
Native works, so it is a possibility to drop pycryptodome.

I'm not sure how much extra work it would be to merge youtube-dl if this changed. I did it back and forth on the existing codebase, and it wasn't that bad. But check history, you'll see.

Over all I'm not sure if I'd like to do anything more with this, good luck.

**Comment by pukkandan** [comment]
I already tried explaining to you in https://github.com/sulyi/yt-dlp/pull/3 that creating separate functions with fallbacks is much easier than wrapping the existing functions. With a native gcm implemented, the only fallback we need is for `cbc` since that is the only one we use to decrypt large quantities of data. If you could recreate a PR with just the gcm implementation, I can separately implement the needed fallback

**Pull request closed without merging**